coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,241 @@
+		defer func() {
+			wut.flushCh <- count
+		}()
+	}
+	if count == 0 {
+		wut.log.Debug(context.Background(), "nothing to flush")
+		return
+	}
+
+	// Set a short-ish timeout for this. We don't want to hang forever.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	// nolint: gocritic // system function
+	authCtx := dbauthz.AsSystemRestricted(ctx)
+	wut.flushLock.Lock()
+	defer wut.flushLock.Unlock()
+	if err := wut.s.BatchUpdateWorkspaceLastUsedAt(authCtx, database.BatchUpdateWorkspaceLastUsedAtParams{
+		LastUsedAt: now,
+		IDs:        ids,
+	}); err != nil {
+		// A single failure to flush is likely not a huge problem. If the workspace is still connected at
+		// the next iteration, either another coderd instance will likely have this data or the CLI
+		// will tell us again that the workspace is in use.
+		wut.flushErrors++
+		if wut.flushErrors > 1 {
+			wut.log.Error(ctx, "multiple failures updating workspaces last_used_at", slog.F("count", count), slog.F("consecutive_errors", wut.flushErrors), slog.Error(err))
+			// TODO: if this keeps failing, it indicates a fundamental problem with the database connection.
+			// How to surface it correctly to admins besides just screaming into the logs?
+		} else {
+			wut.log.Warn(ctx, "failed updating workspaces last_used_at", slog.F("count", count), slog.Error(err))
+		}
+		return
+	}
+	wut.flushErrors = 0
+	wut.log.Info(ctx, "updated workspaces last_used_at", slog.F("count", count), slog.F("now", now))
+}
+
+// Loop periodically flushes every tick.
+// If Loop is called after Close, it will panic. Don't do this.
+func (wut *Tracker) Loop() {
+	// Calling Loop after Close() is an error.
+	select {
+	case <-wut.doneCh:
+		panic("developer error: Loop called after Close")
+	default:
+	}
+	defer func() {
+		wut.log.Debug(context.Background(), "workspace usage tracker loop exited")
+	}()
+	for {
+		select {
+		case <-wut.stopCh:
+			close(wut.doneCh)
+			return
+		case now, ok := <-wut.tickCh:
+			if !ok {
+				return
+			}
+			// NOTE: we do not update last_used_at with the time at which each workspace was added.
+			// Instead, we update with the time of the flush. If the BatchUpdateWorkspacesLastUsedAt
+			// query can be rewritten to update each id with a corresponding last_used_at timestamp
+			// then we could capture the exact usage time of each workspace. For now however, as
+			// we perform this query at a regular interval, the time of the flush is 'close enough'
+			// for the purposes of both dormancy (and for autostop, in future).
+			wut.flush(now.UTC())
+		}
+	}
+}
+
+// Close stops Tracker and returns once Loop has exited.
+// After calling Close(), Loop must not be called.
+func (wut *Tracker) Close() {
+	wut.stopOnce.Do(func() {
+		wut.stopCh <- struct{}{}
+		wut.stopTick()
+		<-wut.doneCh
+	})
+}
+
+// uuidSet is a set of UUIDs. Safe for concurrent usage.
+// The zero value can be used.
+type uuidSet struct {
```

Comment:
How about using a `sync.Map` instead?

Reply:
TBH I'm never sure about the correct use case of sync.Map due to:

https://pkg.go.dev/sync#Map

> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

If you feel strongly that this use case matches the above, I'm fine with it.